frameworklauncher/src/main/java/com/microsoft/frameworklauncher/common/model/ResourceDescriptor.java

```diff
@@ -75,14 +88,56 @@ public void setMemoryMB(Integer memoryMB) {
     this.memoryMB = memoryMB;
   }
 
-  public List<Range> getPortRanges() {
+  public List<ValueRange> getPortRanges() {
     return portRanges;
   }
 
-  public void setPortRanges(List<Range> portRanges) {
+  public void setPortRanges(List<ValueRange> portRanges) {
     this.portRanges = portRanges;
   }
 
+  public Integer getPortNumber() {
+    return portNumber;
+  }
+
+  public void setPortNumber(Integer portNumber) {
+    this.portNumber = portNumber;
+  }
+
+  public Map<String, Ports> getPortDefinitions() {
+    return portDefinitions;
+  }
+
+  public void setPortDefinitions(Map<String, Ports> portDefinitions) throws Exception {
+    this.portDefinitions = portDefinitions;
+    if (portDefinitions == null) {
+      return;
+    }
+
+    // Convert port information from user input format to List<Range> format for AM scheduling.
+    List<ValueRange> portRangeList = new ArrayList<ValueRange>();
+    int portNumber = 0;
+    for (Ports ports : portDefinitions.values()) {
+      if (ports.getStart() != 0) {
+        portRangeList.add(ValueRange.newInstance(ports.getStart(), ports.getStart() + ports.getCount() - 1));
+      } else {
+        portNumber += ports.getCount();
+      }
+    }
+
+    // portNumber and portRangeList are not allow coexistence.
+    // user is not allowed to set "Any" port and "Specified" port in the same task role.
+    // if
+    if (portNumber == 0 && ValueRangeUtils.getValueNumber(portRangeList) > 0) {
+      this.setPortRanges(portRangeList);
+      this.setPortNumber(0);
+    } else if (portNumber > 0 && ValueRangeUtils.getValueNumber(portRangeList) == 0) {
+      this.setPortNumber(portNumber);
```

Comment:
Are you sure this method will be called when deserlize from YARM or JSON?

Reply:
This method will call through JSON, but not call from Yarn. Yarn interface don't have portDefinitions

---
In reply to: [173689119](https://github.com/Microsoft/pai/pull/318#discussion_r173689119) [](ancestors = 173689119)

Reply:
Sorry, I mean deserlize from YAML. 

---
In reply to: [174351035](https://github.com/Microsoft/pai/pull/318#discussion_r174351035) [](ancestors = 174351035,173689119)

Reply:
Pls fix the UT

---
In reply to: [175353577](https://github.com/Microsoft/pai/pull/318#discussion_r175353577) [](ancestors = 175353577,174453415,174351035,173689119)